From: NeilBrown <neilb@suse.de>
Subject: Another fix for race problem with sunrpc cache deferal
Patch-mainline: 2.6.36
References: bnc#632568, bnc#636461

If mountd responds to a request very very quickly, we might find it
the request has been completed before we wait for it, and so not
clean up the sleeper properly.


Acked-by: NeilBrown <neilb@suse.de>
Signed-off-by: Neil Brown <neilb@suse.de>

---
 net/sunrpc/cache.c |    2 ++
 1 file changed, 2 insertions(+)

--- linux-2.6.32-SLE11-SP1.orig/net/sunrpc/cache.c
+++ linux-2.6.32-SLE11-SP1/net/sunrpc/cache.c
@@ -560,6 +560,8 @@ static int cache_defer_req(struct cache_
 	if (!test_bit(CACHE_PENDING, &item->flags)) {
 		/* must have just been validated... */
 		cache_revisit_request(item);
+		if (dreq == &sleeper.handle)
+			wait_for_completion(&sleeper.completion);
 		return -EAGAIN;
 	}
 
